crate/server/src/kms_server.rs

```diff
@@ -185,33 +185,35 @@ pub async fn prepare_kms_server(
     let google_cse_jwks = enable_google_cse.then_some(google_cse::list_jwks_uri());
 
     // Determine if JWT Auth should be used for authentication.
-    let (use_jwt_auth, jwt_config, jwks_manager) = if let Some(jwt_issuer_uri) =
-        &kms_server.params.jwt_issuer_uri
-    {
-        // Prepare all the needed URIs
-        let jwks_uri = JwtAuthConfig::uri(jwt_issuer_uri, kms_server.params.jwks_uri.as_deref());
-        let uris = google_cse_jwks.map_or_else(
-            || vec![jwks_uri.clone()],
-            |mut google_uris| {
-                google_uris.push(jwks_uri.clone());
-                google_uris
-            },
-        );
+    let (use_jwt_auth, jwt_config, jwks_manager) =
+        if let Some(jwt_config) = &kms_server.params.jwt_config {
+            // Prepare all the needed URIs from all the configured Identity Providers
+            let mut all_jwks_uris: Vec<_> = jwt_config
+                .iter()
+                .map(|idp_config| {
+                    JwtAuthConfig::uri(&idp_config.jwt_issuer_uri, Some(&idp_config.jwks_uri))
+                })
+                .collect();
+            // Add the one from google is cse is enabled
+            if let Some(google_cse_jwks) = google_cse_jwks {
+                all_jwks_uris.extend(google_cse_jwks)
+            }
 
-        let jwks_manager = Arc::new(JwksManager::new(uris).await?);
+            let jwks_manager = Arc::new(JwksManager::new(all_jwks_uris).await?);
 
-        (
-            true,
-            Some(Arc::new(JwtConfig {
-                jwt_issuer_uri: jwt_issuer_uri.clone(),
-                jwks: jwks_manager.clone(),
-                jwt_audience: kms_server.params.jwt_audience.clone(),
-            })),
-            Some(jwks_manager),
-        )
-    } else {
-        (false, None, None)
-    };
+            let jwt_configs = jwt_config
+                .iter()
+                .map(|idp_config| JwtConfig {
+                    jwt_issuer_uri: idp_config.jwt_issuer_uri.clone(),
+                    jwks: jwks_manager.clone(),
+                    jwt_audience: idp_config.jwt_audience.clone(),
+                })
+                .collect::<Vec<_>>();
+
+            (true, Some(Arc::new(jwt_configs)), Some(jwks_manager))
+        } else {
+            (false, None, None)
```

Comment:
Your return type couples decision-making (`use_jwt_auth`) with data unwrapping (`jwt_config, jwks_manager`). It may be cleaner to keep them separate:
```suggestion
            Some((Arc::new(jwt_configs), jwks_manager))
        } else {
            None
```

Reply:
The more I think about it, the more I would like this code to look like:
```rust
pub async fn prepare_kms_server(
    kms_server: Arc<KMS>,
    builder: Option<SslAcceptorBuilder>,
) -> KResult<actix_web::dev::Server> {
    let is_google_cse_enabled = kms_server.params.google_cse_kacls_url.is_some();

    let jwks_manager = extract_jwks_manager(&kms_server.params.idp_configs, is_google_cse_enabled);
    let jwt_configs = jwks_manager.as_ref().map(|manager|extract_jwt_configs(&kms_server.params.idp_configs, manager));
    ...
}

Reply:
I put the boolean appart